server/sonar-server/src/main/java/org/sonar/server/computation/scm/ReportScmInfo.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.scm;
+
+import com.google.common.base.Function;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.Immutable;
+import org.sonar.batch.protocol.output.BatchReport;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.FluentIterable.from;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * ScmInfo implementation based on the changeset information from the Report
+ */
+@Immutable
+class ReportScmInfo implements ScmInfo {
+  private final ScmInfo delegate;
+
+  ReportScmInfo(BatchReport.Changesets changesets) {
+    requireNonNull(changesets);
+    this.delegate = convertToScmInfo(changesets);
+  }
+
+  private ScmInfo convertToScmInfo(final BatchReport.Changesets changesets) {
+    return new ScmInfoImpl(
+      from(new IntRangeIterable(changesets.getChangesetIndexByLineCount()))
+        .transform(new LineIndexToChangeset(changesets)));
+  }
+
+  private static Changeset convert(BatchReport.Changesets.Changeset changeset) {
+    checkState(changeset.hasRevision(), "Changeset has no revision");
```

Comment:
I would say "Changeset must have a revision" or "Revision must be set in Changeset"